readthedocs/proxito/views/mixins.py

```diff
@@ -30,52 +36,129 @@ class ServeDocsMixin:
+            storage_path=storage_path,
+            storage_backend=build_media_storage,
+        )
+
+    def _serve_dowload(self, request, project, version, type_):
         """
-        Serve documentation in the way specified by settings.
+        Serve downloadable content for the given version.
+
+        The HTTP header ``Content-Disposition`` is added with the proper
+        filename (e.g. "pip-pypa-io-en-latest.pdf" or "pip-pypi-io-en-v2.0.pdf"
+        or "docs-celeryproject-org-kombu-en-stable.pdf").
+        """
+        storage_path = project.get_storage_path(
+            type_=type_,
+            version_slug=version.slug,
+            version_type=version.type,
+            include_file=True,
+        )
+        self._track_pageview(
+            project=project,
+            path=storage_path,
+            request=request,
+            download=True,
+        )
+        response = self._serve_file(
+            request=request,
+            storage_path=storage_path,
+            storage_backend=build_media_storage,
+        )
+
+        # Set the filename of the download.
+        filename_ext = storage_path.rsplit(".", 1)[-1]
+        domain = unicode_slugify(project.subdomain().replace(".", "-"))
+        if project.is_subproject:
+            filename = f"{domain}-{project.alias}-{project.language}-{version.slug}.{filename_ext}"
+        else:
+            filename = f"{domain}-{project.language}-{version.slug}.{filename_ext}"
+        response["Content-Disposition"] = f"filename={filename}"
+        return response
+
+    def _serve_file(self, request, storage_path, storage_backend):
+        """
+        Serve a file from storage.
 
         Serve from the filesystem if using ``PYTHON_MEDIA``. We definitely
         shouldn't do this in production, but I don't want to force a check for
         ``DEBUG``.
 
-        If ``download`` and ``version_slug`` are passed,
-        the HTTP header ``Content-Disposition`` is added with the proper
-        filename (e.g. "pip-pypa-io-en-latest.pdf" or "pip-pypi-io-en-v2.0.pdf"
-        or "docs-celeryproject-org-kombu-en-stable.pdf")
+        :param storage_path: Path to file to serve.
+        :param storage_backend: Storage backend class from where to serve the file.
         """
-
-        self._track_pageview(
-            project=final_project,
-            path=path,
+        storage_url = self._get_storage_url(
             request=request,
-            download=download,
+            storage_path=storage_path,
+            storage_backend=storage_backend,
         )
         if settings.PYTHON_MEDIA:
-            response = self._serve_file_from_python(request, path, build_media_storage)
+            return self._serve_file_from_python(request, storage_url, storage_backend)
+
+        if storage_backend == build_media_storage:
+            root_path = "proxito"
+        elif storage_backend == staticfiles_storage:
+            root_path = "proxito-static"
```

Comment:
Can't we define this as part of the storage object (e.g. `StorageBackend.root_path`) instead of adding this logic here?